### PR TITLE
HTML comments: compatibility with empty table cell

### LINKED ISSUE #10116: HTML comments: compatibility with empty table cell
## Provide a description of the task

Comment inside empty table cell throws an error after calling `editor.getData()`.

Steps to reproduce:
1. Open `HtmlComment` manual test: http://localhost:8125/ckeditor5-html-support/tests/manual/htmlcomment.html.
2. Call `editor.setData( '<table><tr><td><!-- Comment --></td></tr></table>' )`.
3. Call `editor.getData()`.

Error:

```
Uncaught TypeError: Cannot read property 'name' of undefined
    at Mapper.getModelLength (mapper.js:486)
    at Mapper.findPositionIn (mapper.js:550)
    at Mapper.on.priority (mapper.js:108)
    at Mapper.fire (emittermixin.js:199)
    at Mapper.toViewPosition (mapper.js:335)
    at DowncastDispatcher.<anonymous> (downcasthelpers.js:876)
    at DowncastDispatcher.fire (emittermixin.js:199)
    at DowncastDispatcher.convertMarkerAdd (downcastdispatcher.js:413)
    at DataController.toView (datacontroller.js:260)
    at DataController.stringify (datacontroller.js:213)
```

### PR DESCRIPTION
**Pull request by psmyrek**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Internal (table): Fixed model mappings in table cell if paragraph is bound to its parent. Closes #10116.

### CODE REVIEW

File: packages/ckeditor5-table/src/tableediting.js [hunk 1]
```diff
@@ -159,3 +163,38 @@ export default class TableEditing extends Plugin {
...
+// elements located after this paragraph would point either to a non-existent offset inside `tableCell` (if paragraph is empty), or after
+// first character from paragraph's text. See https://github.com/ckeditor/ckeditor5/issues/10116.
+//
+// <tableCell><paragraph></paragraph>^</tableCell> -> <td>^&nbsp;</td>
+//
+// <tableCell><paragraph>foobar</paragraph>^</tableCell> -> <td>foobar^</td>
+//
+// @param {module:engine/view/view~View} editingView
+// @returns {Function}
+function mapModelPositionToView( editingView ) {
```
Comment (ma2ciek) [inline packages/ckeditor5-table/src/tableediting.js:178]: Maybe `mapTableCellModelPositionToView`?

File: packages/ckeditor5-table/src/tableediting.js [hunk 2]
```diff
@@ -159,3 +163,38 @@ export default class TableEditing extends Plugin {
 		return [ TableUtils ];
 	}
 }
+
+// Creates a mapper callback to adjust model position mappings in a table cell containing a paragraph, which is bound to its parent
+// (to the table cell). Only positions after this paragraph have to be adjusted, because after binding this paragraph to the table cell,
+// elements located after this paragraph would point either to a non-existent offset inside `tableCell` (if paragraph is empty), or after
+// first character from paragraph's text. See https://github.com/ckeditor/ckeditor5/issues/10116.
```
Comment (ma2ciek) [inline packages/ckeditor5-table/src/tableediting.js:170]:
```suggestion
// the first character of the paragraph's text. See https://github.com/ckeditor/ckeditor5/issues/10116.
```

**Comment by ma2ciek** [comment]
LGTM but I'm not an expert in this area. Perhaps @Reinmar or @niegowski could review it as well?

File: packages/ckeditor5-table/src/tableediting.js [hunk 3]
```diff
@@ -159,3 +163,38 @@ export default class TableEditing extends Plugin {
...
+			return;
+		}
+
+		const viewNodeBefore = data.mapper.toViewElement( modelNodeBefore );
+		const viewParent = data.mapper.toViewElement( modelParent );
+
+		if ( viewNodeBefore === viewParent ) {
+			// Since the paragraph has already been bound to its parent, update the current position in the model with paragraph's
+			// max offset, so it points to the place which should normally (in all other cases) be the end position of this paragraph.
+			data.viewPosition = editingView.createPositionAt( viewParent, modelNodeBefore.maxOffset );
```
Comment (niegowski) [inline packages/ckeditor5-table/src/tableediting.js:197]: You should use :

```javascript
data.viewPosition = data.mapper.findPositionIn( viewParent, modelNodeBefore.maxOffset );
```

To find a proper offset in the view element. The current implementation works by accident (by the fact that [https://github.com/ckeditor/ckeditor5/blob/8805829e23ddd010d1f3852f98cfc5c08e4d776f/packages/ckeditor5-engine/src/view/element.js#L653](https://github.com/ckeditor/ckeditor5/blob/8805829e23ddd010d1f3852f98cfc5c08e4d776f/packages/ckeditor5-engine/src/view/element.js#L653) is appending the new child to the list) of children.

Also, this is reordering those elements but I guess it's not in the scope of this PR?

File: packages/ckeditor5-table/src/tableediting.js [hunk 4]
```diff
@@ -120,6 +120,10 @@ export default class TableEditing extends Plugin {
 		// Table heading columns conversion (a change of heading rows requires a reconversion of the whole table).
 		conversion.for( 'editingDowncast' ).add( downcastTableHeadingColumnsChange() );
 
+		// Manually adjust model position mappings in a special case, when a table cell contains a paragraph, which is bound
+		// to its parent (to the table cell).
+		editor.data.mapper.on( 'modelToViewPosition', mapTableCellModelPositionToView( editor.editing.view ) );
```
Comment (niegowski) [inline packages/ckeditor5-table/src/tableediting.js:125]: Please add a comment on why this is needed only in the data pipeline.

**Review by niegowski — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-table/src/tableediting.js (hunk 1 above)
Comment (psmyrek) [reply packages/ckeditor5-table/src/tableediting.js]: Renamed.

File: packages/ckeditor5-table/src/tableediting.js (hunk 2 above)
Comment (psmyrek) [reply packages/ckeditor5-table/src/tableediting.js]: Done.

File: packages/ckeditor5-table/src/tableediting.js (hunk 3 above)
Comment (psmyrek) [reply packages/ckeditor5-table/src/tableediting.js]: Thanks, I've changed it as you suggested.

Tasks related to the reordering of elements are already reported (https://github.com/ckeditor/ckeditor5/issues/10118, https://github.com/ckeditor/ckeditor5/issues/10119). This PR is only to fix the editor crash when there was a comment in the empty table cell.

File: packages/ckeditor5-table/src/tableediting.js (hunk 4 above)
Comment (psmyrek) [reply packages/ckeditor5-table/src/tableediting.js]: Added.

**Pull request merged**